docs/content/guides/migrations/from-step-launchers-to-pipes.mdx

```diff
@@ -0,0 +1,333 @@
+
+FROM scratch AS export
+
+COPY --from=builder /output/venv.pex /venv.pex
+```
+
+The build can be launched with:
+
+```shell
+DOCKER_BUILDKIT=1 docker build --output type=local,dest=./output .
+```
+
+Then, upload the produced `output/venv.pix` file to an S3 bucket:
+
+```shell
+aws s3 cp output/venv.pex s3://your-bucket/venv.pex
+```
+
+Upload your scripts to S3 as well.
+
+Finally, use the `--files` and `spark.pyspark.python` options to specify the path to the PEX file in the `spark-submit` command:
+
+```shell
+spark-submit ... --files s3://your-bucket/venv.pex --conf spark.pyspark.python=./venv.pex
+```
+
+This process can be automated with a CI/CD pipeline. You can learn more about alternative ways to prepare your PySpark runtime environment in [PySpark documentation](https://spark.apache.org/docs/latest/api/python/user_guide/python_packaging.html#python-package-management).
+
+It's also possible to run Java or Scala Spark jobs with Dagster Pipes, but currently there is no official Pipes implementation for these languages. Therefore, forwarding Dagster events from these jobs is not yet supported officially (although it can be done with some custom code).
+
+### **2. Update your Dagster code**
+
+The goal is to keep the same observability and orchestration features while moving compute to an external script. Suppose you have existing code using Step Launchers similar to this:
+
+```python file=/guides/migrations/from_step_launchers_to_pipes/old_code.py
+from typing import Any
+
+from dagster_aws.emr import emr_pyspark_step_launcher
+from my_lib import MyPysparkIOManager, calculate_metric, get_data_frame
+from pyspark.sql import DataFrame
+
+import dagster as dg
+
+# the upstream asset will serve as an example of writing a Spark DataFrame
+
+
+@dg.asset(io_manager_key="pyspark_io_manager")
+def upstream(pyspark_step_launcher: dg.ResourceParam[Any]) -> DataFrame:
+    return get_data_frame()
+
+
+# the downstream asset will serve as an example of reading a Spark DataFrame
+# and logging metadata to Dagster
+
+
+@dg.asset(io_manager_key="pyspark_io_manager")
+def downstream(
+    context: dg.AssetExecutionContext,
+    upstream: DataFrame,
+    pyspark_step_launcher: dg.ResourceParam[Any],
+) -> None:
+    my_metric = calculate_metric(upstream)
+    context.add_output_metadata({"my_metric": my_metric})
+    return
+
+
+definitions = dg.Definitions(
+    assets=[upstream, downstream],
+    resources={
+        "pyspark_step_launcher": emr_pyspark_step_launcher,
+        "pyspark_io_manager": MyPysparkIOManager(),
+    },
+)
+```
+
+The corresponding Pipes code will be split in 2 pieces: the Dagster part and the PySpark job part.
+
+Let's start with the PySpark job. The upstream asset will invoke the following script:
+
+```python file=/guides/migrations/from_step_launchers_to_pipes/upstream_asset_script.py
+import os
+
+import boto3
+from dagster_pipes import PipesS3MessageWriter, open_dagster_pipes
+from my_lib import get_data_frame
```

Comment:
Can we simplify this by instantiating a SparkSession, and not reference a fictional `my_lib`?

Reply:
I think the point of this is to clearly show how the business logic is being moved to the external script, I would rather keep it this way